File: leetcode/0695-max-area-of-island/solution.cpp
int dfs(int row, int col, vector<vector<int>>& grid, vector<vector<bool>>& seen) {
    const int m = grid.size();
    const int n = grid[0].size();

    seen[row][col] = true;

    vector<vector<int>> directions = { {1, 0}, {0, 1}, {-1, 0}, {0, -1} };
    int island_size = 0;

    for (int i = 0; i < directions.size(); i++) {
        int new_row = row + directions[i][0];
        int new_col = col + directions[i][1];

        if (new_row < 0 || new_row >= m || new_col < 0 || new_col >= n) {
            continue;
        }

        if (grid[new_row][new_col] == 1 && !seen[new_row][new_col]) {
            island_size += dfs(new_row, new_col, grid, seen);
        }
    }

    return island_size + 1;
}

int maxAreaOfIsland(vector<vector<int>>& grid) {
    const int m = grid.size();
    const int n = grid[0].size();

    vector<vector<bool>> seen(m, vector<bool>(n, false));

    int max_area = 0;

    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (grid[i][j] == 1 && !seen[i][j]) {
                max_area = max(max_area, dfs(i, j, grid, seen));
            }
        }
    }

    return max_area;
}

File: leetcode/0695-max-area-of-island/solution.ts
function maxAreaOfIsland(grid: number[][]): number {
  const directions = [
    [1, 0],
    [0, 1],
    [-1, 0],
    [0, -1]
  ];

  const m = grid.length;
  const n = grid[0].length;

  const seen: boolean[][] = Array.from({ length: m }, () =>
    Array.from<boolean>({ length: n }).fill(false)
  );

  let maxArea = 0;

  function dfs(row: number, col: number): number {
    seen[row][col] = true;

    const isValid = (_row: number, _col: number) => {
      if (_row < 0 || _row >= m) return false;
      if (_col < 0 || _col >= n) return false;
      return true;
    };

    let neighbourIslandSize = 0;

    for (let i = 0; i < directions.length; i++) {
      const _row = row + directions[i][0];
      const _col = col + directions[i][1];

      if (isValid(_row, _col) && grid[_row][_col] && !seen[_row][_col]) {
        neighbourIslandSize += dfs(_row, _col);
      }
    }

    return neighbourIslandSize + 1;
  }

  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] && !seen[i][j]) {
        maxArea = Math.max(maxArea, dfs(i, j));
      }
    }
  }

  return maxArea;
}
